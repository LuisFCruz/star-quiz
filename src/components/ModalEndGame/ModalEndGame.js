import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  FormControl,
  Input,
  InputLabel,
  Typography,
  withStyles,
} from '@material-ui/core';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';

import { savePlayer } from '../../apis/starwars-api';

const styles = {
  score: {
    textAlign: 'center',
    fontWeight: '700',
    fontSize: '18px',
  },
  fieldControl: {
    marginTop: '10px',
    display: 'block',
    width: '100%'
  }
}

export class ModalEndGame extends Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    finished: PropTypes.bool,
    score: PropTypes.number,
    dispatch: PropTypes.func,
  }

  constructor() {
    super();

    this.state = {
      name: '',
      email: '',
      error: false,
      open: true,
    }
  }

  handleClose = () => {
    this.setState({ open: false });
  }

  handleConfirm = () => {
    const { name, email } = this.state;
    const { score } = this.props;
    const error = !name;
    if (!error) {
      savePlayer({name, email, score});
      this.handleClose();
    }

    this.setState({ error });
  }


  setName = ({target: { value = '' }}) => {
    this.setState({ name: value });
  }

  setEmail = ({target: { value = '' }}) => {
    this.setState({ email: value });
  }

  render() {
    const { classes, finished, score } = this.props;
    const { error, open } = this.state

    return (
      <Dialog onClose={this.handleClose} open={finished && open}>
        <DialogTitle>Game over!</DialogTitle>
        <DialogContent>
          <Typography className={classes.score}>{score} points</Typography>
          <Typography>Save your score!</Typography>
          <FormControl className={classes.fieldControl} error={error}>
            <InputLabel>Name</InputLabel>
            <Input onChange={this.setName} />
          </FormControl>
          <FormControl className={classes.fieldControl}>
            <InputLabel>Email</InputLabel>
            <Input onChange={this.setEmail} />
          </FormControl>
        </DialogContent>
        <DialogActions>
          <Button onClick={this.handleClose} color="primary">
            Cancel
          </Button>
          <Button onClick={this.handleConfirm} color="primary">
            Save
          </Button>
        </DialogActions>
      </Dialog>
    )
  }
}

const mapStateToProps = (state) => {
  const { finished, score } = state;
  return { finished, score };
}

export default connect(mapStateToProps)(withStyles(styles)(ModalEndGame));
