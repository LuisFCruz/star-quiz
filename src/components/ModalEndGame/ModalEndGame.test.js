import React from 'react';
import { shallow } from 'enzyme';
import { Input, Button } from '@material-ui/core';

import { ModalEndGame } from './ModalEndGame';

describe('<ModalEndGame />', () => {
  const props = {
		classes: {},
		finished: false,
	};
 let wrapper;
 let component;
 
	beforeEach(() => {
		wrapper = shallow(<ModalEndGame {...props} />);
		component = wrapper.instance();
	});

	test('smoke test', () => {
		expect(ModalEndGame).toBeDefined();
		expect(component).toBeInstanceOf(ModalEndGame);
	});

  test('render', () => {
    expect(wrapper).toMatchSnapshot();
	});

	test('should have call setName() on input changed', () => {
		const handleSpy = jest.fn();
    component.setName = handleSpy;
    component.forceUpdate();

    const inputs = wrapper.find(Input);
		expect(inputs).toHaveLength(2);

    inputs.first().simulate('change', { target: { value: 'value changed'  } });
    expect(handleSpy).toHaveBeenCalled();
	});
	
	test('should have set name in state', () => {
		component.setName({ target: { value: 'User' } });
		expect(wrapper.state('name')).toBe('User');
	});

	test('should have call setEmail() on input changed', () => {
		const handleSpy = jest.fn();
    component.setEmail = handleSpy;
    component.forceUpdate();

    const inputs = wrapper.find(Input);
		expect(inputs).toHaveLength(2);

    inputs.last().simulate('change', { target: { value: 'value changed'  } });
    expect(handleSpy).toHaveBeenCalled();
	});

	test('should have set email in state', () => {
		component.setEmail({ target: { value: 'user@user.com' } });
		expect(wrapper.state('email')).toBe('user@user.com');
	});

	test('should have call handleClose() on click in close button', () => {
		const handleSpy = jest.fn();
    component.handleClose = handleSpy;
    component.forceUpdate();

    const buttons = wrapper.find(Button);
		expect(buttons).toHaveLength(2);

    buttons.first().simulate('click');
    expect(handleSpy).toHaveBeenCalled();
	});

	test('should have set open in state', () => {
		component.handleClose();
		expect(wrapper.state('open')).toBeFalsy();
	});

	test('should have call handleConfirm() on click in save button', () => {
		const handleSpy = jest.fn();
    component.handleConfirm = handleSpy;
    component.forceUpdate();

    const buttons = wrapper.find(Button);
		expect(buttons).toHaveLength(2);

    buttons.last().simulate('click');
    expect(handleSpy).toHaveBeenCalled();
	});
});
