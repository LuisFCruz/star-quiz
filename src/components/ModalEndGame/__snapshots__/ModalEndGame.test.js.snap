// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ModalEndGame /> render 1`] = `
<WithStyles(Dialog)
  onClose={[Function]}
  open={false}
>
  <WithStyles(DialogTitle)>
    Game over!
  </WithStyles(DialogTitle)>
  <WithStyles(DialogContent)>
    <WithStyles(Typography)>
       points
    </WithStyles(Typography)>
    <WithStyles(Typography)>
      Save your score!
    </WithStyles(Typography)>
    <WithStyles(FormControl)
      error={false}
    >
      <WithStyles(WithFormControlContext(InputLabel))>
        Name
      </WithStyles(WithFormControlContext(InputLabel))>
      <WithStyles(Input)
        onChange={[Function]}
      />
    </WithStyles(FormControl)>
    <WithStyles(FormControl)>
      <WithStyles(WithFormControlContext(InputLabel))>
        Email
      </WithStyles(WithFormControlContext(InputLabel))>
      <WithStyles(Input)
        onChange={[Function]}
      />
    </WithStyles(FormControl)>
  </WithStyles(DialogContent)>
  <WithStyles(DialogActions)>
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Cancel
    </WithStyles(Button)>
    <WithStyles(Button)
      color="primary"
      onClick={[Function]}
    >
      Save
    </WithStyles(Button)>
  </WithStyles(DialogActions)>
</WithStyles(Dialog)>
`;
